eoe/man/man1/split.1
'\"macro stdmacro
.if n .pH g1.split @(#)split	41.7 of 5/26/91
.\" Copyright 1991 UNIX System Laboratories, Inc.
.\" Copyright 1989, 1990 AT&T
.nr X
.if \nX=0 .ds x} split 1 "Directory and File Management Utilities" "\&"
.if \nX=1 .ds x} split 1 "Directory and File Management Utilities"
.if \nX=2 .ds x} split 1 "" "\&"
.if \nX=3 .ds x} split "" "" "\&"
.TH \*(x}
.SH NAME
\f4split\f1 \- split a file into pieces
.SH SYNOPSIS
\f4split\f1 [\f4\-l\f2line_count\f1] [\f4\-a\f2suffix_length\f1] 
[\f2file\f1 [\f2name\f1]]
.br
\f4split\f1 \f4\-b\f2n\f1[k|m] [\f4\-a\f2suffix_length\f1] 
[\f2file\f1 [\f2name\f1]]
.br
\f4split\f1
[
\f4\-\f2line_count\f1
]
[ \f2file\f1 [ \f2name\f1 ] ]
.SH DESCRIPTION
\f4split\fP
reads an input
.I file
and writes
it in one or more output files. The default size of each
output file is 1000 lines. The size of the output files can
be modified by \f4\-b\f1 and \f4\-l\f1 options. Each
output file is created with a unique suffix. The suffix 
consists of exactly \f2suffix_length\f1 lower-case letters from
locale. The letters of the suffix are used as if they were
a base-26 digit system, with the first suffix to be created 
consisting of all a characters, the second with a b replacing
the last a, and so sorth, until a name of all z characters is
created. By default, The names of the output files are x,
followed by a two-character suffix from the character set as 
described above, starting with aa, ab, ac, and so forth, and
continuing until the suffix zz, for a maximum of 676 files.
.PP
If the number of files required exceeds the maximum allowed
by the suffix length provided, such that the last allowable
file would be larger than the requested size, \f4split\f1
will fail after creating the last file with a valid suffix;
it will not delete the files it created with valid suffixes.
If the file limit is not exceeded, the last file created will 
contain the remainder of the input file, and may be smaller than
the requested size.
.PP
\f2file\f1 is the pathname of the file to be split. If no file
is given or \f2file\f1 is \f4\-\f1, the standard input will be
used. \f2name\f1 is the prefix to be used for each of the files
resulting from the split operation. If not given, x will be used 
as the prefix of the output files.
.SH OPTIONS
.TP
\f4\-a\f2suffix_length\f1
Use \f2suffix_length\f1 letters to form the suffix portion of the 
filenames of the split file. Default suffix length is 2. If
the sum of \f2name\f1 and \f2suffix_length\f1 option-argument
would create a filename exceeding NAME_MAX bytes, \f4split\f1
will exit with a diagnostic message and no files will be
created.
.TP
\f4\-b\f2n\f1[k|m}
Specify the output file size in bytes \f2n\f1. With k specified, an output
file is n*1024 bytes in size;with m specified, a file is n*1048576 
bytes in size.
.TP
\f4\-l\f2line_count\f1 or \f4\-\f2line_count\f1
Specify the number of lines in each output file. \f2line_count\f1
is an unsigned decimal integer. The default is 1000. If the input
does not end with a newline character, the partial line will be
included in the last ouput file.
.SH FILES
.TP
\f4/usr/lib/locale/\f2locale\f4/LC_MESSAGES/uxdfm\f1
language-specific message file (see \f4LANG\fP on \f4environ\f1(5)
.SH SEE ALSO
\f4bfs\fP(1), \f4csplit\fP(1),
\f4statvfs\fP(2)
.\"	@(#)split.1	6.2 of 9/2/83
.Ee


eoe/cmd/split/split.c
/*	Copyright (c) 1990, 1991 UNIX System Laboratories, Inc.	*/
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989, 1990 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF     	*/
/*	UNIX System Laboratories, Inc.                     	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

/*	Portions Copyright (c) 1988, Sun Microsystems, Inc.	*/
/*	All Rights Reserved. 					*/

/* #ident	"@(#)split:split.c	1.6.1.2" */
#ident	"$Header: /proj/irix6.5.7m/isms/eoe/cmd/split/RCS/split.c,v 1.7 1998/08/03 15:42:23 morimoto Exp $"

#include <stdio.h>
#include <sys/types.h>
#include <sys/statvfs.h>
#include <locale.h>
#include <pfmt.h>
#include <string.h>
#include <errno.h>

unsigned count = 1000;        /* default # of lines in each output file */
int     suffix_length = 2;    /* default suffix length */    
int     nbytes;               /* bytes in each output file if -b is specified*/
int     setlines = 0;
int     setbytes = 0;         /* default split a file according to line # */
int	fnumber;              /* actual # of output file */
char	fname[100];           /* name of output file */
char	head[1024];           /* output file directory */
char	*ifil;                /* input filename */
char	*ofil;                /* output full filename */
char	*tail;                /* output file basename */
char	*last;
FILE	*is;                  /* input file stream */
FILE	*os;                  /* output file stream */
int     *base;

main(argc, argv)
char *argv[];
{
    register i, j, c, f;
    int iflg = 0;
    struct statvfs64 stbuf;
    int maxfile;
    int delimit = 0;

    (void)setlocale(LC_ALL, "");
    (void)setcat("uxdfm");
    (void)setlabel("UX:split");

    for (argc--, argv++; argc > 0; argc--, argv++)
	if(**argv == '-' && delimit == 0){
	    if(argv[0][1] == '-') 
	      delimit = 1;
	    else if(isdigit(argv[0][1])) 
	      count = atoi(argv[0]+1);
	    else 
	      switch(argv[0][1]){
		case '\0':
		  iflg = 1;
		  break;

		case 'a':
		  if(argv[0][2] == '\0'){
		      argc--; argv++;
		      suffix_length = atoi(*argv);
		  } else
		    suffix_length = atoi(argv[0]+2);
		  if(suffix_length <= 0){
		      pfmt(stderr, MM_ERROR, ":2:Incorrect usage\n");
		      pfmt(stderr, MM_ERROR, ":65:suffix_length should be positive.\n");
		      exit(1);
		  }
		  break;
		  
		case 'l':
		  if(setbytes == 1){
		      pfmt(stderr, MM_ERROR, ":2:Incorrect usage\n");
		      pfmt(stderr, MM_ERROR, ":67:-b and -l shouldn't be used together.\n");
		      exit(1);
		  }
		  setlines = 1;
		  if(argv[0][2] == '\0'){
		      argc--; argv++;
		      count = atoi(*argv);
		  }
		  else
		    count = atoi(argv[0]+2);
		  break;

		case 'b':
		  if(setlines == 1){
		      pfmt(stderr, MM_ERROR, ":2:Incorrect usage\n");
		      pfmt(stderr, MM_ERROR, ":67:-b and -l shouldn't be used together.\n");
		      exit(1);
		  }
		  setbytes = 1;
		  if(argv[0][2] == '\0'){
		      argc--; argv++;
		      nbytes = atoi(*argv);
		  }
		  else
		    nbytes = atoi(argv[0]+2);
		  if(nbytes <= 0){
		      pfmt(stderr, MM_ERROR, ":2:Incorrect usage\n");
		      pfmt(stderr, MM_ERROR, ":68:Number of bytes should be positive.\n");
		      exit(1);
		  }
		  if(argv[0][strlen(argv[0])-1] == 'k')
		    nbytes *= 1024;
		  else if (argv[0][strlen(argv[0])-1] == 'm')
		    nbytes *= 1048576;
		  else if (!isdigit(argv[0][strlen(argv[0])-1]))
		    usage();
		  break;
		  
		default:
		  usage();
	      }
	    
	    if(count <= 0){
		pfmt(stderr, MM_ERROR, ":2:Incorrect usage\n");
		pfmt(stderr, MM_ERROR, ":66:line_count should be positive.\n");
		exit(1);
	    }
	} else if(iflg)         /* process input and output filename */
	  ofil = *argv;
	else {
	    ifil = *argv;
	    iflg = 2;
	} /* finish command line processing */
	  
    if(iflg != 2)
      is = stdin;
    else
      if((is=fopen(ifil,"r")) == NULL) {
	  pfmt(stderr, MM_ERROR, ":3:Cannot open %s: %s\n",
	       ifil, strerror(errno));
	  exit(1);
      }

    if(ofil == 0)
      ofil = "x";
    else {
	if ((tail = strrchr(ofil, '/')) == NULL) {
	    tail = ofil;
	    getcwd(head, sizeof(head));
	}
	else {
	    tail++;
	    strcpy(head, ofil);
	    last = strrchr(head, '/');
	    *++last = '\0';
	}
		
	if (statvfs64(head, &stbuf) < 0) {
	    pfmt(stderr, MM_ERROR, ":56:%s: %s\n", head, strerror(errno));
	    exit(1);
	}

	if (strlen(tail) > (stbuf.f_namemax-suffix_length) ) {
	    pfmt(stderr, MM_ERROR, ":57:More than %d characters in output file name\n",
		 stbuf.f_namemax-suffix_length);
	    exit(1);
	}
    }

    if((base = (int *)malloc(suffix_length * sizeof(int))) == NULL){
	pfmt(stderr, MM_ERROR, ":69:Cannot allocate enough memory.\n");
	exit(1);
    }
    base[0] = 1;
    for (i=1; i<suffix_length; i++)
      base[i] = base[i-1]*26;

    maxfile = base[suffix_length-1]*26;
    
    for(;;){
	f = 1;
	if(!setbytes)        /* set line limit */
	  for(i=0; i<count; i++)
	    do {
		c = getc(is);
		if(c == EOF) {
		    if(f == 0)
		      fclose(os);
		    exit(0);
		}
		if(f) { /* need a new output file */
		    int temp = fnumber;
		    for(f=0; ofil[f]; f++)
		      fname[f] = ofil[f];
		    for(j=suffix_length-1; j>=0; j--){
			fname[f++] = temp/base[j] + 'a';
			temp %= base[j];
		    }
		    fname[f] = '\0';
		    if(++fnumber > maxfile) {
			pfmt(stderr, MM_ERROR,
			     ":58:More output files needed, aborting split\n");
			exit(1);
		    }
		    if((os=fopen(fname,"w")) == NULL) {
			pfmt(stderr, MM_ERROR,
			     ":12:Cannot create %s: %s\n",
			     fname, strerror(errno));
			exit(1);
		    }
		    f = 0;
		}
		putc(c, os);
	    } while(c != '\n');
	else             /* set bytes limit */
	  for(i=0; i<nbytes; i++){
	      c = getc(is);
	      if(c == EOF) {
		  if(f == 0)
		    fclose(os);
		  exit(0);
	      }
	      if(f) { /* need a new output file */
		  int temp = fnumber;
		  for(f=0; ofil[f]; f++)
		    fname[f] = ofil[f];
		  for(j=suffix_length-1; j>=0; j--){
		      fname[f++] = temp/base[j] + 'a';
		      temp %= base[j];
		  }
		  fname[f] = '\0';
		  if(++fnumber > maxfile) {
		      pfmt(stderr, MM_ERROR,
			   ":58:More output files needed, aborting split\n");
		      exit(1);
		  }
		  if((os=fopen(fname,"w")) == NULL) {
		      pfmt(stderr, MM_ERROR,
			   ":12:Cannot create %s: %s\n",
			   fname, strerror(errno));
		      exit(1);
		  }
		  f = 0;
	      }
	      putc(c, os);
	  }
	fclose(os);
    }
}

usage()
{
    pfmt(stderr, MM_ERROR, ":2:Incorrect usage\n");
    pfmt(stderr, MM_ACTION, ":55:Usage: split [-l line_count][-a suffix_length][file[name]]\nor\nsplit -b n[k|m][-a suffix_length][file[name]]\n");
    exit(1);
}


